README.md
# San ANode Utils

Util Functions for San [ANode](https://github.com/baidu/san/blob/master/doc/anode.md). 和 [San](https://github.com/baidu/san) [ANode](https://github.com/baidu/san/blob/master/doc/anode.md) 相关的功能函数集。

[![NPM version](http://img.shields.io/npm/v/san-anode-utils.svg?style=flat-square)](https://npmjs.org/package/san-anode-utils)
[![License](https://img.shields.io/github/license/ecomfe/san-anode-utils.svg?style=flat-square)](https://npmjs.org/package/san-anode-utils)
[![CircleCI](https://circleci.com/gh/ecomfe/san-anode-utils.svg?style=svg)](https://circleci.com/gh/ecomfe/san-anode-utils)


## 版本说明

由于 [San](https://github.com/baidu/san) 在不同的 minor 版本号之间，[ANode](https://github.com/baidu/san/blob/master/doc/anode.md) 会有细微差别。所以，san-anode-utils 的版本号与 [San](https://github.com/baidu/san) 版本号二位兼容。也就是说，san-anode-utils `3.9.x` 版本支持 [San](https://github.com/baidu/san) `3.9.x`。

## 安装

```
$ npm i --save san-anode-utils
```

## Properties

### ExprType

表达式类型枚举常量。包含表达式见下方示例。

`Object`

```js
ExprType = {
    STRING,
    NUMBER,
    BOOL,
    ACCESSOR,
    INTERP,
    CALL,
    TEXT,
    BINARY,
    UNARY,
    TERTIARY,
    OBJECT,
    ARRAY,
    NULL
};
```

## Functions

### parseTemplate

`parseTemplate({string}template): {ANode}`

将模板字符串解析成 [ANode](https://github.com/baidu/san/blob/master/doc/anode.md) 对象。

```js
const aNodeUtils = require('san-anode-utils');

let aNode = aNodeUtils.parseTemplate('<p>Hello {{name}}</p>');
/* aNode
{
    "directives": {},
    "props": [],
    "events": [],
    "children": [
        {
            "directives": {},
            "props": [],
            "events": [],
            "children": [
                {
                    "textExpr": {
                        "type": 7,
                        "segs": [
                            {
                                "type": 1,
                                "value": "Hello "
                            },
                            {
                                "type": 5,
                                "expr": {
                                    "type": 4,
                                    "paths": [
                                        {
                                            "type": 1,
                                            "value": "name"
                                        }
                                    ]
                                },
                                "filters": []
                            }
                        ]
                    }
                }
            ],
            "tagName": "p"
        }
    ]
}
*/
```

### parseComponentTemplate

将组件的模板解析成 [ANode](https://github.com/baidu/san/blob/master/doc/anode.md) 对象。与直接调用 `parseTemplate` 不同，`parseComponentTemplate` 会自动抽取第一个子元素作为组件根元素，为其附加 `id/style/class` 的逻辑，其行为与运行时组件编译完全相同。

`parseComponentTemplate({Function}ComponentClass): {ANode}`

```js
const aNodeUtils = require('san-anode-utils');
const san = require('san');

const MyComponent = san.defineComponent({
    template: '<p>Hello {{name}}</p>'
});
let aNode = aNodeUtils.parseComponentTemplate(MyComponent);
/* aNode
{
    "directives": {},
    "props": [
        {
            "name": "class",
            "expr": {
                "type": 5,
                "expr": {
                    "type": 4,
                    "paths": [
                        {
                            "type": 1,
                            "value": "class"
                        }
                    ]
                },
                "filters": [
                    {
                        "type": 6,
                        "args": [],
                        "name": {
                            "type": 4,
                            "paths": [
                                {
                                    "type": 1,
                                    "value": "_class"
                                }
                            ]
                        }
                    }
                ]
            }
        },
        {
            "name": "style",
            "expr": {
                "type": 5,
                "expr": {
                    "type": 4,
                    "paths": [
                        {
                            "type": 1,
                            "value": "style"
                        }
                    ]
                },
                "filters": [
                    {
                        "type": 6,
                        "args": [],
                        "name": {
                            "type": 4,
                            "paths": [
                                {
                                    "type": 1,
                                    "value": "_style"
                                }
                            ]
                        }
                    }
                ]
            }
        },
        {
            "name": "id",
            "expr": {
                "type": 4,
                "paths": [
                    {
                        "type": 1,
                        "value": "id"
                    }
                ]
            }
        }
    ],
    "events": [],
    "children": [
        {
            "textExpr": {
                "type": 7,
                "segs": [
                    {
                        "type": 1,
                        "value": "Hello "
                    },
                    {
                        "type": 5,
                        "expr": {
                            "type": 4,
                            "paths": [
                                {
                                    "type": 1,
                                    "value": "name"
                                }
                            ]
                        },
                        "filters": []
                    }
                ]
            }
        }
    ],
    "tagName": "p"
}
*/
```

### parseExpr

将表达式源码解析成 [ANode](https://github.com/baidu/san/blob/master/doc/anode.md) 对象。

`parseExpr({string}source): {ANode}`

```js
const aNodeUtils = require('san-anode-utils');

let aNode = aNodeUtils.parseExpr('num + 1');
/* aNode
{
    "type": 8,
    "operator": 43,
    "segs": [
        {
            "type": 4,
            "paths": [
                {
                    "type": 1,
                    "value": "num"
                }
            ]
        },
        {
            "type": 2,
            "value": 1
        }
    ]
}
*/
```

### pack

将 [ANode](https://github.com/baidu/san/blob/master/doc/anode.md) 压缩成 [APack](https://github.com/baidu/san/blob/master/doc/anode-pack.md)。

`pack({ANode}source): {Array}`

```js
const aNodeUtils = require('san-anode-utils');

let aNode = aNodeUtils.parseTemplate('<p>Hello {{name}}</p>');
let aPack = aNodeUtils.pack(aNode.children[0]);
/* aPack
[
    1,
    "p",
    1,
    undefined,
    9,
    undefined,
    2,
    3,
    "Hello ",
    7,
    undefined,
    6,
    1,
    3,
    "name",
    undefined
]
*/
```

### pack.stringify

将 [APack](https://github.com/baidu/san/blob/master/doc/anode.md) 转换成字符串。

`pack.stringify({ANode}source): {Array}`

```js
const aNodeUtils = require('san-anode-utils');

let aNode = aNodeUtils.parseTemplate('<p>Hello {{name}}</p>');
let aPack = aNodeUtils.pack(aNode.children[0]);
let aPackString = aNodeUtils.pack.stringify(aPack);
/* aPackString
[1,"p",1,,9,,2,3,"Hello ",7,,6,1,3,"name",]
*/
```

### unpack

将 [APack](https://github.com/baidu/san/blob/master/doc/anode-pack.md) 解压缩成 [ANode](https://github.com/baidu/san/blob/master/doc/anode.md)。

`unpack({Array}aPack): {ANode}`

```js
const aNodeUtils = require('san-anode-utils');

let aNode = aNodeUtils.unpack([1,"p",1,,9,,2,3,"Hello ",7,,6,1,3,"name",]);
/* aNode
{
    "directives": {},
    "props": [],
    "events": [],
    "children": [
        {
            "textExpr": {
                "type": 7,
                "segs": [
                    {
                        "type": 1,
                        "value": "Hello "
                    },
                    {
                        "type": 5,
                        "expr": {
                            "type": 4,
                            "paths": [
                                {
                                    "type": 1,
                                    "value": "name"
                                }
                            ]
                        },
                        "filters": []
                    }
                ]
            }
        }
    ],
    "tagName": "p"
}
*/
```

## License

san-anode-utils is [MIT licensed](./LICENSE).


CHANGELOG.md
ChangeLog
========

3.9.1
-------

+ 【优化】- pack.stringify 方法支持自动压缩 ANode

test/perf.js
const fs = require('fs');
const path = require('path');
const utils = require('../index');


const packDir = path.resolve(__dirname, '../node_modules/san-anode-cases/pack');

let template = fs.readFileSync(`${packDir}/composite.tpl`, 'UTF-8');
let an = utils.parseTemplate(template).children[0];

let packed = utils.pack(an);
const times = 1000;

console.time();
for (var i = 0; i < times; i++) {
    utils.parseTemplate(template).children[0];
}
console.timeEnd();


console.time();
for (var i = 0; i < times; i++) {
    utils.unpack(packed);
}
console.timeEnd();


test/pack.test.js
const fs = require('fs');
const path = require('path');
const utils = require('../index');

const packDir = path.resolve(__dirname, '../node_modules/san-anode-cases/pack');
const files = fs.readdirSync(packDir);

files.forEach(file => {
    file = path.parse(file);
    if (file.ext === '.tpl') {
        test(`pack ${file.name}`, () => {
            const tpl = fs.readFileSync(`${packDir}/${file.base}`, 'UTF-8');
            const result = fs.readFileSync(`${packDir}/${file.name}.apack`, 'UTF-8'); 

            let aNode = utils.parseTemplate(tpl).children[0];
            expect(utils.pack.stringify(aNode)).toBe(result);
        });
    }
});


test/unpack.test.js
const fs = require('fs');
const path = require('path');
const utils = require('../index');


const packDir = path.resolve(__dirname, '../node_modules/san-anode-cases/pack');
const files = fs.readdirSync(packDir);

files.forEach(file => {
    file = path.parse(file);
    if (file.ext === '.tpl') {
        test(`unpack ${file.name}`, () => {
            const tpl = fs.readFileSync(`${packDir}/${file.base}`, 'UTF-8');
            const result = packResult(fs.readFileSync(`${packDir}/${file.name}.apack`, 'UTF-8'));
            
            expectEqual(utils.parseTemplate(tpl).children[0], utils.unpack(result));
        });
    }
});

function packResult(source) {
    return (new Function(`return ${source}`))();
}

function expectEqual(a, b) {
    var type = typeof a;
    expect(type).toBe(typeof b);

    if (a === null) {
        expect(b === null).toBeTruthy();
    }
    else {
        switch (type) {
            case 'boolean':
            case 'string':
            case 'number':
                expect(b === a).toBeTruthy();
                break;

            case 'undefined':
                expect(b == null).toBeTruthy();
                break;

            case 'object':
                for (var i in a) {
                    if (a.hasOwnProperty(i) && a[i] != null) {
                        expect(b.hasOwnProperty(i)).toBeTruthy();
                    }
                }
            
                for (var i in b) {
                    if (b.hasOwnProperty(i) && b[i] != null) {
                        expect(a.hasOwnProperty(i)).toBeTruthy();
                    }
                }
            
                for (var i in a) {
                    expectEqual(a[i], b[i]);
                }
                break;

            default:
                if (a instanceof Array) {
                    expect(b instanceof Array).toBeTruthy();
                    expect(a.length).toBe(b.length);

                    if (b instanceof Array && a.length === b.length) {
                        for (var i = 0; i < a.length; i++) {
                            expectEqual(a[i], b[i]);
                        }
                    }
                }
        }
    }
}


src/compress.js
/**
 * Copyright (c) Baidu Inc. All rights reserved.
 *
 * This source code is licensed under the MIT license.
 * See LICENSE file in the project root for license information.
 *
 * @file ANode 压缩
 */

var ExprType = require('san').ExprType;

function compressTemplate(aNode) {
    if (aNode.textExpr) {
        return [void(0)].concat(compressExpr(aNode.textExpr));
    }

    var result = [1, aNode.tagName, aNode.children.length || void(0)];
    
    // compress children
    for (var i = 0; i < aNode.children.length; i++) {
        result = result.concat(compressTemplate(aNode.children[i]));
    }

    // compress prop
    for (var i = 0; i < aNode.props.length; i++) {
        var prop = aNode.props[i];
        result = result.concat(
            prop.x ? 34 : (prop.noValue ? 33 : 2),
            prop.name,
            compressExpr(prop.expr)
        );
    }

    // compress events
    for (var i = 0; i < aNode.events.length; i++) {
        var event = aNode.events[i];
        result = result.concat(35, compressExpr(event.expr));
        
        var modifierLen = 0;
        var modifierResult = [0];
        for (var key in event.modifier) {
            modifierLen++;
            modifierResult.push(key);
        }
        modifierResult[0] = modifierLen;

        if (modifierLen) {
            result = result.concat(modifierResult);
        }
        else {
            result.push(void(0));
        }
    }

    // compress directives
    for (var key in aNode.directives) {
        var directive = aNode.directives[key];
        switch (key) {
            case 'if':
                result = result.concat(38, compressExpr(directive.value));
                if (directive.elses instanceof Array) {
                    for (var i = 0; i < directive.elses.length; i++) {
                        result = result.concat(compressTemplate(directive.elses[i]));
                    }
                }
                else {
                    result.push(void(0));
                }
                break;
                
            case 'else':
                result.push(40);
                break;
                
            case 'elif':
                result = result.concat(39, compressExpr(directive.value));
                break;
                
            case 'for':
                result = result.concat(
                    37, 
                    compressExpr(directive.value),
                    directive.item,
                    directive.index || void(0),
                    directive.trackByRaw || void(0)
                );
                break;
                
            case 'html':
                result = result.concat(43, compressExpr(directive.value));
                break;
                
            case 'bind':
                result = result.concat(42, compressExpr(directive.value));
                break;
                
            case 'ref':
                result = result.concat(41, compressExpr(directive.value));
                break;
                
            case 'transition':
                result = result.concat(44, compressExpr(directive.value));
                break;
        }
    }

    // compress vars
    if (aNode.vars) {
        for (var i = 0; i < aNode.vars.length; i++) {
            var varItem = aNode.vars[i];
            result = result.concat(36, varItem.name, compressExpr(varItem.expr));
        }
    }

    return result;
}

function compressExpr(expr) {
    var result;

    switch (expr.type) {
        case ExprType.STRING:
            return [3, expr.value];

        case ExprType.NUMBER:
            return [4, expr.value];

        case ExprType.BOOL:
            return [5, expr.value ? 1 : void(0)];

        case ExprType.NULL:
            return [19];

        case ExprType.ACCESSOR:
            result = [6, expr.paths.length];
            for (var i = 0; i < expr.paths.length; i++) {
                result = result.concat(compressExpr(expr.paths[i]));
            }
            break;

        case ExprType.INTERP:
            result = [7].concat(
                compressExpr(expr.expr),
                expr.original ? 1 : void(0),
                expr.filters.length || void(0)
            );
            for (var i = 0; i < expr.filters.length; i++) {
                result = result.concat(compressExpr(expr.filters[i]));
            }
            break;

        case ExprType.CALL:
            result = [8].concat(compressExpr(expr.name));
            result.push(expr.args.length || void(0));
            for (var i = 0; i < expr.args.length; i++) {
                result = result.concat(compressExpr(expr.args[i]));
            }
            break;

        case ExprType.TEXT:
            result = [9, expr.original ? 1 : void(0), expr.segs.length || void(0)];
            for (var i = 0; i < expr.segs.length; i++) {
                result = result.concat(compressExpr(expr.segs[i]));
            }
            break;
        
        case ExprType.BINARY:
            result = [10, expr.operator, expr.segs.length || void(0)];
            for (var i = 0; i < expr.segs.length; i++) {
                result = result.concat(compressExpr(expr.segs[i]));
            }
            break;
        
        case ExprType.UNARY:
            result = [11, expr.operator].concat(compressExpr(expr.expr));
            break;

        case ExprType.TERTIARY:
            result = [12];
            for (var i = 0; i < expr.segs.length; i++) {
                result = result.concat(compressExpr(expr.segs[i]));
            }
            break;

        case ExprType.OBJECT:
            result = [13, expr.items.length || void(0)];
            for (var i = 0; i < expr.items.length; i++) {
                var item = expr.items[i];
                result = result.concat(
                    item.spread ? 15 : 14,
                    compressExpr(item.name),
                    compressExpr(item.expr)
                );
            }
            break;

        case ExprType.ARRAY:
            result = [16, expr.items.length || void(0)];
            for (var i = 0; i < expr.items.length; i++) {
                var item = expr.items[i];
                result = result.concat(
                    item.spread ? 18 : 17,
                    compressExpr(item.expr)
                );
            }
            break;
    }

    return result || [];
}

module.exports = exports = function (aNode) {
    return compressTemplate(aNode);
};

exports.stringify = function (compressed) {
    return '['
        + compressed.map(function (item) {
            if (item == null) {
                return item;
            }

            return JSON.stringify(item);
        }).join(',')
        + ']';
};

src/pack.js
/**
 * Copyright (c) Baidu Inc. All rights reserved.
 *
 * This source code is licensed under the MIT license.
 * See LICENSE file in the project root for license information.
 *
 * @file ANode 压缩
 */

var ExprType = require('san').ExprType;

function packTemplateNode(aNode) {
    if (aNode.textExpr) {
        return [void(0)].concat(packExpr(aNode.textExpr));
    }

    var len = aNode.children.length + aNode.props.length + aNode.events.length;
    if (aNode.vars) {
        len += aNode.vars.length;
    }
    for (var key in aNode.directives) {
        switch (key) {
            case 'if':
            case 'else':
            case 'elif':
            case 'bind':
            case 'html':
            case 'for':
            case 'ref':
            case 'transition':
                len++;
        }
    }

    var result = [1, aNode.tagName, len || void(0)];

    // pack prop
    for (var i = 0; i < aNode.props.length; i++) {
        var prop = aNode.props[i];
        result = result.concat(
            prop.x ? 34 : (prop.noValue ? 33 : 2),
            prop.name,
            packExpr(prop.expr)
        );
    }

    // pack events
    for (var i = 0; i < aNode.events.length; i++) {
        var event = aNode.events[i];
        result = result.concat(35, event.name, packExpr(event.expr));

        var modifierLen = 0;
        var modifierResult = [0];
        for (var key in event.modifier) {
            modifierLen++;
            modifierResult.push(key);
        }
        modifierResult[0] = modifierLen;

        if (modifierLen) {
            result = result.concat(modifierResult);
        }
        else {
            result.push(void(0));
        }
    }

    // pack directives
    for (var key in aNode.directives) {
        var directive = aNode.directives[key];
        switch (key) {
            case 'if':
                result = result.concat(38, packExpr(directive.value));
                break;

            case 'else':
                result.push(40);
                break;

            case 'elif':
                result = result.concat(39, packExpr(directive.value));
                break;

            case 'for':
                result = result.concat(
                    37,
                    directive.item,
                    directive.index || void(0),
                    directive.trackByRaw || void(0),
                    packExpr(directive.value)
                );
                break;

            case 'html':
                result = result.concat(43, packExpr(directive.value));
                break;

            case 'bind':
                result = result.concat(42, packExpr(directive.value));
                break;

            case 'ref':
                result = result.concat(41, packExpr(directive.value));
                break;

            case 'transition':
                result = result.concat(44, packExpr(directive.value));
                break;
        }
    }

    // pack vars
    if (aNode.vars) {
        for (var i = 0; i < aNode.vars.length; i++) {
            var varItem = aNode.vars[i];
            result = result.concat(36, varItem.name, packExpr(varItem.expr));
        }
    }

    // pack children
    for (var i = 0; i < aNode.children.length; i++) {
        result = result.concat(packTemplateNode(aNode.children[i]));
    }

    if (aNode.directives['if']) {
        if (aNode.elses instanceof Array) {
            result.push(aNode.elses.length || void(0));
            for (var i = 0; i < aNode.elses.length; i++) {
                result = result.concat(packTemplateNode(aNode.elses[i]));
            }
        }
        else {
            result.push(void(0));
        }
    }

    return result;
}

function packExpr(expr) {
    var result;

    switch (expr.type) {
        case ExprType.STRING:
            return [3, expr.value];

        case ExprType.NUMBER:
            return [4, expr.value];

        case ExprType.BOOL:
            return [5, expr.value ? 1 : void(0)];

        case ExprType.NULL:
            return [19];

        case ExprType.ACCESSOR:
            result = [6, expr.paths.length];
            for (var i = 0; i < expr.paths.length; i++) {
                result = result.concat(packExpr(expr.paths[i]));
            }
            break;

        case ExprType.INTERP:
            result = [7].concat(
                expr.original ? 1 : void(0),
                packExpr(expr.expr),
                expr.filters.length || void(0)
            );
            for (var i = 0; i < expr.filters.length; i++) {
                result = result.concat(packExpr(expr.filters[i]));
            }
            break;

        case ExprType.CALL:
            result = [8].concat(packExpr(expr.name));
            result.push(expr.args.length || void(0));
            for (var i = 0; i < expr.args.length; i++) {
                result = result.concat(packExpr(expr.args[i]));
            }
            break;

        case ExprType.TEXT:
            result = [9, expr.original ? 1 : void(0), expr.segs.length || void(0)];
            for (var i = 0; i < expr.segs.length; i++) {
                result = result.concat(packExpr(expr.segs[i]));
            }
            break;

        case ExprType.BINARY:
            result = [10, expr.operator];
            for (var i = 0; i < expr.segs.length; i++) {
                result = result.concat(packExpr(expr.segs[i]));
            }
            break;

        case ExprType.UNARY:
            result = [11, expr.operator].concat(packExpr(expr.expr));
            break;

        case ExprType.TERTIARY:
            result = [12];
            for (var i = 0; i < expr.segs.length; i++) {
                result = result.concat(packExpr(expr.segs[i]));
            }
            break;

        case ExprType.OBJECT:
            result = [13, expr.items.length || void(0)];
            for (var i = 0; i < expr.items.length; i++) {
                var item = expr.items[i];
                if (item.spread) {
                    result = result.concat(15, packExpr(item.expr));
                }
                else {
                    result = result.concat(14, packExpr(item.name), packExpr(item.expr));
                }
            }
            break;

        case ExprType.ARRAY:
            result = [16, expr.items.length || void(0)];
            for (var i = 0; i < expr.items.length; i++) {
                var item = expr.items[i];
                result = result.concat(
                    item.spread ? 18 : 17,
                    packExpr(item.expr)
                );
            }
            break;
    }

    return result || [];
}

module.exports = exports = function (aNode) {
    return packTemplateNode(aNode);
};

exports.stringify = function (packed) {
    if (typeof packed === 'object' && !(packed instanceof Array)) {
        packed = packTemplateNode(packed);
    }

    if (packed instanceof Array) {
        return '['
            + packed.map(function (item) {
                if (item == null) {
                    return item;
                }

                return JSON.stringify(item);
            }).join(',')
            + ']';
    }

    return '';
};